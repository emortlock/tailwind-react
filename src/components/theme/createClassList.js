import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
