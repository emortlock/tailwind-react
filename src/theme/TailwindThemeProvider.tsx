import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
