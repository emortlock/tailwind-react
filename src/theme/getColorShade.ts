const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
