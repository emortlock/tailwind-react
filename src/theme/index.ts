export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
