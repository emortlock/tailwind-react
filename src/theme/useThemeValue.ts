import { tailwindPropToClassName } from 'tailwind-react-primitives'

export const useThemeValue = (
  prefix: string,
  value: unknown,
  userClassNames = '',
) => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
