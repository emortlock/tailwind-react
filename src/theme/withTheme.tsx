import type { Theme } from './TailwindTheme'

import React from 'react'

import { TailwindTheme } from './TailwindTheme'

export interface WithTheme {
  theme: Theme
}

export const withTheme = <P extends React.PropsWithChildren<{ theme: Theme }>>(
  Component: React.FC<P>,
) => {
  const WithTheme: React.FC<Omit<P, 'theme'>> = (props) => (
    <TailwindTheme.Consumer>
      {(theme) => {
        // @ts-ignore FIXME: Constraint error
        return <Component {...props} theme={theme} />
      }}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}
