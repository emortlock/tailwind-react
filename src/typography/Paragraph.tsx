import type { TextProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'

import React from 'react'
import { RawText as Text, getAsArray } from 'tailwind-react-primitives'

import { withTheme } from '../theme'

export interface ParagraphProps<E extends HTMLElement = HTMLParagraphElement>
  extends WithTheme,
    Omit<TextProps<E>, 'size'> {
  lead?: boolean
  size?: number
}

export const RawParagraph = <E extends HTMLElement = HTMLParagraphElement>({
  theme,
  children,
  as = 'p',
  size,
  lead,
  text,
  ...rest
}: ParagraphProps<E>) => (
  <Text<E>
    as={as}
    text={
      [
        (size || lead) &&
          theme.text.size.body[
            (lead ? theme.text.size.body.length : size) ?? 2 - 1
          ],
        ...getAsArray(text),
      ].filter(Boolean) as string[]
    }
    m={{ b: theme.spacing.md }}
    {...rest}
  >
    {children}
  </Text>
)

export const Paragraph = withTheme<ParagraphProps>(RawParagraph)
