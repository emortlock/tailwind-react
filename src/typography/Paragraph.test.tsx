import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawParagraph as Paragraph } from './Paragraph'
import { defaultTheme } from '../theme'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = render(
    <Paragraph {...props}>Lorem ipsum dolor sit amet</Paragraph>,
  )

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('Paragraph', () => {
  it('renders matching snapshot', () => {
    const [{ asFragment }] = setup({ as: 'p' })

    expect(asFragment()).toMatchSnapshot()
  })
})
