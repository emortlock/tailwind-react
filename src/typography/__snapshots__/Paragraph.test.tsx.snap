// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paragraph renders matching snapshot 1`] = `
<DocumentFragment>
  <p
    class="leading-normal mb-4"
  >
    Lorem ipsum dolor sit amet
  </p>
</DocumentFragment>
`;
